src/delta/delta.c

```diff
@@ -207,102 +208,99 @@ static int found_override(const char *top, const char *bottom) {
         _cleanup_free_ char *unit = NULL;
-        _cleanup_free_ char *path = NULL;
         _cleanup_strv_free_ char **list = NULL;
-        char **file;
-        char *c;
+        char **file, *c, *path;
         int r;
 
         assert(!endswith(drop, "/"));
 
-        path = strjoin(toppath, "/", drop, NULL);
-        if (!path)
-                return -ENOMEM;
+        log_debug("Looking at %s", toppath);
 
-        log_debug("Looking at %s", path);
-
-        unit = strdup(drop);
-        if (!unit)
+        path = strdup(toppath);
+        if (!path)
                 return -ENOMEM;
 
-        c = strrchr(unit, '.');
+        c = strrchr(path, '.');
         if (!c)
                 return -EINVAL;
         *c = 0;
 
-        r = get_files_in_directory(path, &list);
+        r = get_files_in_directory(toppath, &list);
         if (r < 0)
-                return log_error_errno(r, "Failed to enumerate %s: %m", path);
+                return log_error_errno(r, "Failed to enumerate %s: %m", toppath);
+
 
         STRV_FOREACH(file, list) {
-                Hashmap *h;
                 int k;
-                char *p;
-                char *d;
+                Hashmap *drops;
+                char **d, *tmp = NULL, *key;
 
                 if (!endswith(*file, ".conf"))
                         continue;
 
-                p = strjoin(path, "/", *file, NULL);
-                if (!p)
-                        return -ENOMEM;
-                d = p + strlen(toppath) + 1;
-
-                log_debug("Adding at top: %s %s %s", d, draw_special_char(DRAW_ARROW), p);
-                k = hashmap_put(top, d, p);
-                if (k >= 0) {
-                        p = strdup(p);
-                        if (!p)
-                                return -ENOMEM;
-                        d = p + strlen(toppath) + 1;
-                } else if (k != -EEXIST) {
-                        free(p);
+                k = unit_name_template(basename(path), &tmp);
+                if (k < 0 && k != -EINVAL) {
+                        free(path);
                         return k;
                 }
+                key = tmp ? tmp : basename(path);
 
-                log_debug("Adding at bottom: %s %s %s", d, draw_special_char(DRAW_ARROW), p);
-                free(hashmap_remove(bottom, d));
-                k = hashmap_put(bottom, d, p);
-                if (k < 0) {
-                        free(p);
-                        return k;
-                }
+                drops = hashmap_get(bottom, key);
+                if (!drops) {
+                        _cleanup_free_ char *p;
 
-                h = hashmap_get(drops, unit);
-                if (!h) {
-                        h = hashmap_new(&string_hash_ops);
-                        if (!h)
+                        drops = hashmap_new(&string_hash_ops);
+                        if (!drops) {
+                                free(tmp);
```

Comment:
Again, it's used as a key, it's not supposed to be freed when it goes out of scope.